### PR TITLE
Protect CSI driver node pods to avoid storage workload scheduling

### PR DESCRIPTION
**Pull request by alikdell**
# Description
The CSI node driver when not running on a WN (worker node). the CSI controller should check and taint that node so that pods are not scheduled on that node.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| #https://github.com/dell/csm/issues/145  |

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] Unit test added
- [x] Test run integration test

time="2022-05-19T14:00:25-04:00" level=info msg="Node-mode test finished"
--- PASS: TestNodeMode (1.84s)
=== RUN   TestMapEqualsMap
--- PASS: TestMapEqualsMap (0.00s)
=== RUN   TestPowerFlexShortCheck
time="2022-05-19T14:00:25-04:00" level=info msg="Skipping short integration test. To enable short integration test: export RESILIENCY_SHORT_INT_TEST=true"
--- PASS: TestPowerFlexShortCheck (0.00s)
=== RUN   TestUnityShortCheck
time="2022-05-19T14:00:25-04:00" level=info msg="Skipping short integration test. To enable short integration test: export RESILIENCY_SHORT_INT_TEST=true"
--- PASS: TestUnityShortCheck (0.00s)
=== RUN   TestPowerFlexShortIntegration
time="2022-05-19T14:00:25-04:00" level=info msg="Skipping integration test. To enable integration test: export RESILIENCY_SHORT_INT_TEST=true"
--- PASS: TestPowerFlexShortIntegration (0.00s)
=== RUN   TestUnityShortIntegration
time="2022-05-19T14:00:25-04:00" level=info msg="Skipping integration test. To enable integration test: export RESILIENCY_SHORT_INT_TEST=true"
--- PASS: TestUnityShortIntegration (0.00s)
PASS
coverage: 92.9% of statements
status 0
ok      podmon/internal/monitor 8.292s  coverage: 92.9% of statements

### CODE REVIEW

File: internal/monitor/monitor_steps_test.go [hunk 1]
```diff
@@ -1047,6 +1047,40 @@ func (f *feature) iCallGetPodAffinityLabels() error {
...
+	}
+	f.err = f.podmonMonitor.controllerModeDriverPodHandler(f.pod, eventType)
+	if f.pod2 != nil {
+		f.podmonMonitor.controllerModeDriverPodHandler(f.pod2, eventType)
+	}
+	return nil
+}
+
+func (f *feature) theIsTainted(node string) error {
+	return nil
```
Comment (rbo54) [inline internal/monitor/monitor_steps_test.go:1081]: Pass the boolean in here and then you can check both the addition of the taint as well as the removal of the taint.

File: internal/monitor/node.go [hunk 2]
```diff
@@ -291,6 +298,7 @@ func (pm *PodMonitorType) nodeModeCleanupPods(node *v1.Node) bool {
 
 		// Check containers to make sure they're not running. This uses the containerInfos map obtained above.
 		pod := podInfo.Pod
+		namespace, name := splitPodKey(podKey)
```
Comment (rbo54) [inline internal/monitor/node.go:301]: I'm not sure why you moved this from below the for loop; looks like it doesn't make any difference. Doesn't really matter.

**Review by rbo54 — CHANGES_REQUESTED** [review]
A few small changes I think would make it better but overall very good work!

File: internal/monitor/node.go (hunk 2 above)
Comment (alikdell) [reply internal/monitor/node.go:301]: I moved that when I thought I needed to check in this function also if pod is driver node pod, forgot to move it down back to original place.

File: internal/monitor/controller.go [hunk 3]
```diff
@@ -634,3 +642,72 @@ func mapEqualsMap(map1, map2 map[string]string) bool {
...
+	if pod.Spec.NodeName != "" {
+		log.Debugf("Getting node %s", pod.Spec.NodeName)
+		node, err := K8sAPI.GetNode(ctx, pod.Spec.NodeName)
+		if err != nil {
+			log.Errorf("GetNode failed: %s: %s", pod.Spec.NodeName, err)
+		} else {
+			// Determine pod status
+			ready := false
+			initialized := true
+			conditions := pod.Status.Conditions
```
[the remark this answers is not in the document]
Comment (alikdell) [reply internal/monitor/controller.go]: fixed

File: internal/monitor/monitor.go [hunk 4]
```diff
@@ -174,6 +179,7 @@ func podMonitorHandler(eventType watch.EventType, object interface{}) error {
 	pm := &PodMonitor
 	switch PodMonitor.Mode {
 	case "controller":
+		// driver-namespace == pod.spec.namespace call different function.
```
[the remark this answers is not in the document]
Comment (alikdell) [reply internal/monitor/monitor.go]: fixed

File: internal/monitor/node.go (hunk 2 above)
Comment (alikdell) [reply internal/monitor/node.go]: reverted

**Review by rbo54 — APPROVED** [review]
Thanks for the changes Alik! Approved.

**Pull request merged**